### PR TITLE
feat: add support for custom compile options in torch_xla.compile and…

### PR DESCRIPTION
**Pull request by sshonTT**
… PJRT backend

Issue description : https://github.com/pytorch/xla/issues/9555

This change introduces the ability to pass custom compile options from Python down to the PJRT backend, allowing users to fine-tune XLA compilation behavior without modifying core code. Key changes:
* Python API
    * Added custom_compile_options parameter to torch_xla.compile for passing compile-time options as a dict (supports bool, float, int, and str values).
    * Added torch_xla.set_custom_compile_options() utility for setting compile options globally.
    * Added internal binding _XLAC._set_custom_compile_options().
* C++ Runtime
    * Added SetCustomCompileOptions() virtual method to ComputationClient and implemented it in PjRtComputationClient.
    * PjRtComputationClient now stores custom_compile_options_ and injects them into xla::CompileOptions.env_option_overrides during compilation.
    * Options are stringified before being passed to XLA for compatibility. Motivation: This enables advanced users to pass through backend-specific tuning flags (e.g., enabling experimental optimizations, toggling partitioning strategies) without hardcoding them, improving flexibility for research and debugging workflows.

### CODE REVIEW

**Comment by sshonTT** [comment]
@qihqi Thank you for your review! I realized there was a lint issue and have fixed it. Would you mind re-running the tests?

**Comment by qihqi** [comment]
hi @sshonTT would you rebase to latest HEAD? the 2 CI issue should be fixed

**Comment by sshonTT** [comment]
> hi @sshonTT would you rebase to latest HEAD? the 2 CI issue should be fixed

@qihqi thank you for the info. I had rebased so could you re-run test?

**Comment by sshonTT** [comment]
hi @qihqi , gentle reminder.

**Comment by sshonTT** [comment]
Hi @zhanyong-wan — not sure if this is the right tag. I’m waiting on @qihqi’s follow-up (he previously approved). I rebased per his suggestion and want to confirm CI is green. Could you trigger CI for this PR and, if you have time, give it a review? Thanks!

File: torch_xla/csrc/runtime/computation_client.h [hunk 1]
```diff
@@ -447,6 +447,9 @@ class ComputationClient {
   // after the last ':' character of the device string.
   static int64_t GetDeviceOrdinal(const std::string& device);
 
+  virtual void SetCustomCompileOptions(
```
Comment (zhanyong-wan) [inline torch_xla/csrc/runtime/computation_client.h:454]: Document the format of the argument?

File: torch_xla/torch_xla.py [hunk 2]
```diff
@@ -116,6 +116,7 @@ def compile(
     full_graph: Optional[bool] = False,
     name: Optional[str] = None,
     max_different_graphs: Optional[int] = None,
+    custom_compile_options: Optional[dict] = None,
```
Comment (zhanyong-wan) [inline torch_xla/torch_xla.py]: Optional[dict[str, Any]]

File: torch_xla/torch_xla.py [hunk 3]
```diff
@@ -136,6 +137,8 @@ def compile(
       max_different_graphs (Optional[int]): number of different traced graphs of the given
         model/function that we are allowed to have. An error will be raised in case this limit
         is exceeded.
+      custom_compile_options (Optional[dict]): A dictionary of custom compile options to be set.
+        The keys are strings and the values can be of type bool, float, int, or str.
```
Comment (zhanyong-wan) [inline torch_xla/torch_xla.py]: Please document that the keys are XLA compiler flag names.

File: torch_xla/torch_xla.py [hunk 4]
```diff
@@ -215,6 +218,8 @@ def _compile():
       torch_xla._XLAC._set_use_eager_mode(saved_eager_mode_status)
       torch_xla._XLAC._set_current_graph_name(saved_current_graph_name)
 
+  if custom_compile_options is not None and len(custom_compile_options) > 0:
```
Comment (zhanyong-wan) [inline torch_xla/torch_xla.py]: Typically we just do
```
custom_compile_options = custom_compile_options or {}
```
once at the beginning of the function to normalize this to a dict, and then use it as a dict everywhere, without the need to validate it again.

File: torch_xla/torch_xla.py [hunk 5]
```diff
@@ -264,3 +269,15 @@ def launch(
     fn(xu.getenv_as(xenv.LOCAL_RANK, int), *args)
   else:
     xmp.spawn(fn, args=args, nprocs=nprocs, start_method=start_method)
+
+
+def set_custom_compile_options(options: Optional[dict] = None,):
```
Comment (zhanyong-wan) [inline torch_xla/torch_xla.py]: We should not give a default value to options. The caller should always provide a dict when calling this function (it's confusing what the behavior of `set_custom_compile_options()` without an argument is).

```
def set_custom_compile_options(options: dict[str, Any]) -> None:
```

File: torch_xla/csrc/runtime/computation_client.h (hunk 1 above)
Comment (zhanyong-wan) [inline torch_xla/csrc/runtime/computation_client.h:454]: Document what happens if this is called multiple times? (Are the options overwriting existing options or appending to them? If the latter, the function should be named Add*)

Comment (sshonTT) [reply torch_xla/csrc/runtime/computation_client.h:454]: Thanks! I added explicit docs: the keys are XLA compiler flag names (env option overrides) and values are stringified. I also referenced this in the Python docstring and clarified the overwrite semantics in the C++ header.

File: torch_xla/torch_xla.py (hunk 2 above)
Comment (sshonTT) [reply torch_xla/torch_xla.py]: Updated the signature to Optional[dict[str, Any]]

File: torch_xla/torch_xla.py (hunk 5 above)
Comment (sshonTT) [reply torch_xla/torch_xla.py]: Dropped the default None and now require options: dict[str, Any]. Passing {} clears the options.

File: torch_xla/torch_xla.py (hunk 4 above)
Comment (sshonTT) [reply torch_xla/torch_xla.py]: I think it is still needed because setting custom compile options is working as overwrite. If I pass empty dict here, compile option would be cleared which is not expected.

File: torch_xla/torch_xla.py [hunk 6]
```diff
@@ -136,6 +137,9 @@ def compile(
       max_different_graphs (Optional[int]): number of different traced graphs of the given
         model/function that we are allowed to have. An error will be raised in case this limit
         is exceeded.
+      custom_compile_options (Optional[dict[str, Any]]): XLA **compiler flag overrides**
```
Comment (zhanyong-wan) [inline torch_xla/torch_xla.py]: Document how None and an empty dict are different?

File: torch_xla/torch_xla.py [hunk 7]
```diff
@@ -215,6 +219,9 @@ def _compile():
       torch_xla._XLAC._set_use_eager_mode(saved_eager_mode_status)
       torch_xla._XLAC._set_current_graph_name(saved_current_graph_name)
 
+  custom_compile_options = custom_compile_options or {}
+  if len(custom_compile_options) > 0:
```
Comment (zhanyong-wan) [inline torch_xla/torch_xla.py]: This means there's no way for the user to clear the custom compile options?

File: torch_xla/torch_xla.py (hunk 6 above)
Comment (sshonTT) [reply torch_xla/torch_xla.py]: They are doing same behaviour (no-op).

Comment (sshonTT) [reply torch_xla/torch_xla.py]: Add doc about this.

File: torch_xla/torch_xla.py (hunk 7 above)
Comment (sshonTT) [reply torch_xla/torch_xla.py]: I’m exposing torch_xla.set_custom_compile_options for that purpose.
Users can clear the options by calling this API with an empty dict {}.

Comment (zhanyong-wan) [reply torch_xla/torch_xla.py]: The behavior of compile() should be consistent with set_custom_compile_options(). If given an empty dict, both should do the same thing. Otherwise the API is very confusing and error-prone.

I suggest:

- keep set_custom_compile_options() unchanged,
- treat None as no-op here,
- treat {} as "clear options" here.

File: torch_xla/torch_xla.py [hunk 8]
```diff
@@ -264,3 +272,14 @@ def launch(
...
+
+def set_custom_compile_options(options: dict[str, Any]) -> None:
+  """Set XLA **compiler flag overrides** (env option overrides) for compilation.
+
+  Args:
+    options: Dict mapping XLA flag names to values. Values may be bool/float/int/str;
+      they will be stringified before being passed to XLA.
+      Pass an empty dict `{}` to clear previously set options.
+  """
+  torch_xla._XLAC._set_custom_compile_options(options or {})
```
Comment (zhanyong-wan) [inline torch_xla/torch_xla.py]: `or {}` is unnecessary here. `options` cannot be None according to the type annotation.

**Review by zhanyong-wan — COMMENTED** [review]
In the future, could you avoid git rebase or git merge during a code review? Otherwise it's hard to track the changes. Thanks!

File: torch_xla/torch_xla.py (hunk 8 above)
Comment (sshonTT) [reply torch_xla/torch_xla.py]: deleted it.

File: torch_xla/csrc/runtime/pjrt_computation_client.cpp [hunk 9]
```diff
@@ -554,13 +554,18 @@ std::vector<ComputationClient::ComputationPtr> PjRtComputationClient::Compile(
 
   for (auto& instance : instances) {
     xla::CompileOptions compile_options;
+    for (auto& option : custom_compile_options_) {
```
Comment (zhanyong-wan) [inline torch_xla/csrc/runtime/pjrt_computation_client.cpp]: Make the reference const and use structured binding:

`const auto& [name, value] : ...`

**Review by zhanyong-wan — APPROVED** [review]

**Comment by sshonTT** [comment]
hi @qihqi, I see https://github.com/pytorch/xla/pull/9634 fixes build failure in ci, so I rebased it.

**Comment by sshonTT** [comment]
I noticed the efforts in #9650 and #9655 regarding the current build failure. Hopefully they will resolve the issue. I’ve rebased once again.

**Comment by sshonTT** [comment]
I just noticed that `GetComputationClientOrDie` API is deprecated after https://github.com/pytorch/xla/pull/9620/files which is my fault, had to care about it. I just remove `GetComputationClientOrDie`.

**Comment by sshonTT** [comment]
@zhanyong-wan , @qihqi Could you re-trigger this test? I had checked on my local, and want to see if it is working.

**Comment by sshonTT** [comment]
Thanks for running CI! Can this be merged in?

**Comment by zhanyong-wan** [comment]
Yes, please resolve the comments and merge.
